### PR TITLE
Replicated Sharing Tensor: Multiplication, Matrix multiplication, and Convolution

### PR DESCRIPTION
**Pull request by abogaziah**
## Description
The implementation of the rest of the operations of replicated shared tensor that's necessary to implement falcon protocol.
1- Multiplication: to multiplay a secret x = [x1,x2,x3] by a secret y = [y1,y2,y3] : 

- Parties locally compute z1 = x1y1+x2y1+x1y2, z2 = x2y2+x3y2+x2y3 and z3 = x3y3 +x1y3 +x3y1.
At the end of this, z1,z2, and z3 form a 3-out-of-3 secret sharing of z

- Each party adds 3of3 correlated random number (alpha i) to zi where i is the index of the worker.

- Parties perform resharing by sending a copy of zi to party i+1

2- Matrix multiplication: multiplication algorithm works with matrix multiplication, just by replacing multiplication operator with matrix multiplication

3-  Convolution: convolution is done by converting to one matrix multiplication as explained [here](https://cs231n.github.io/convolutional-networks/)  using [torch unfold](https://discuss.pytorch.org/t/custom-convolution-dot-product/14992/27) 

4- Applied black box principle, i.e. the class should take in a tensor and a set of workers and output a tensor, all other details should be abstracted inside the class, I did that by removing shares_map from the parameters and change other detail around it  

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests
______________________________________________________
- [x] Multiplication #3900 
- [x] Matrix multiplication #3901 
- [x] convolution #3901

### CODE REVIEW

File: syft/frameworks/torch/mpc/przs.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 import torch
 import syft
 
-RING_SIZE = 2 ** 64
+RING_SIZE = 2 ** 5
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/przs.py:8]: Q: should this be 5?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 2]
```diff
@@ -122,12 +112,120 @@ def private_linear_operation(self, secret, operator):
...
+    def private_unfold(secret, kernel_size, padding):
+        secret = secret.apply_to_shares(torch.Tensor.double)
+        secret = secret.apply_to_shares(
+            torch.nn.functional.unfold, kernel_size=kernel_size, padding=padding
+        )
+        secret = secret.apply_to_shares(torch.Tensor.long)
+        return secret
+
+    @staticmethod
+    def __switch_public_private(value, public_function, private_function, *args, **kwargs):
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:219]: I really like this function :D

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 3]
```diff
@@ -11,29 +11,32 @@ def __init__(
...
         return workers
 
     def generate_shares(self, plain_text, number_of_shares=3):
         shares = []
+        plain_text.long()
         for _ in range(number_of_shares - 1):
-            shares.append(torch.tensor(random.randrange(self.ring_size)))
-        shares.append(torch.tensor((plain_text - sum(shares)) % self.ring_size))
+            shares.append(torch.randint(high=self.ring_size // 2, size=plain_text.shape))
+        shares.append((plain_text - sum(shares) % self.ring_size))
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:39]: Q: ```)``` should be after ```(plain_text - sum(shares))```

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 4]
```diff
@@ -146,6 +244,37 @@ def get_shares_map(self):
...
+        shares_map = {
+            player: (
+                function(shares_map[player][0], *args, **kwargs),
+                function(shares_map[player][1], *args, **kwargs),
+            )
+            for player in players
+        }
+        return ReplicatedSharingTensor(shares_map)
+
+    def apply_to_shares_(self, function, *args, **kwargs):
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:263]: Q: Is this used?

File: syft/frameworks/torch/mpc/przs.py (hunk 1 above)
Comment (abogaziah) [reply syft/frameworks/torch/mpc/przs.py:8]: it should be less than or equal the ring size of RST

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 2 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:219]: yeah me too :D it saved a lot of code repetition

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 3 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:39]: yes it should

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 4 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:263]: No, I thought I will need it for convolution, but I didn't have to do any in-place operations, should I remove it? 
I think I should remove it, but I'm like 99% sure I'll need it in the future and I'll regret removing it 😂

Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py]: removed it

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 5]
```diff
@@ -1,39 +1,54 @@
...
+    def share_secret(self, plain_text, workers):
         number_of_shares = len(workers)
         workers = self.__arrange_workers(list(workers))
-        shares = self.generate_shares(secret, number_of_shares)
+        shares = self.generate_shares(plain_text, number_of_shares)
         shares_map = self.__distribute_shares(workers, shares)
-        self.child = shares_map
-        return self
+        PRZS.setup(workers)
+        return shares_map
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:33]: Q: Shouldn't we return ```self``` here?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 6]
```diff
@@ -1,39 +1,54 @@
...
 
 
 class ReplicatedSharingTensor(AbstractTensor):
     def __init__(
-        self, shares_map=None, owner=None, id=None, tags=None, description=None,
+        self, plain_text=None, players=None, owner=None, id=None, tags=None, description=None,
     ):
         super().__init__(id=id, owner=owner, tags=tags, description=description)
+        self.ring_size = 2 ** 32
+        shares_map = self.__generate_shares_map(plain_text, players)
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:14]: Q: Shouldn't we do this only if we have ```plain_text != None```?
(maybe we can have two ways in which we construct RSS, one when we send the plaintext to the constructor and one using the ```share``` method)

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 7]
```diff
@@ -1,39 +1,54 @@
...
         self.child = shares_map
-        self.ring_size = 2 ** 5
 
-    def share_secret(self, secret, workers):
+    def __generate_shares_map(self, plain_text, players):
+        if plain_text is None or players is None:
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:18]: Q: Shouldn't we better raise an exception here? (maybe something like ```BadValue```)

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 8]
```diff
@@ -1,39 +1,54 @@
...
+            raise ValueError("you must provide 3 players")
+        me = syft.hook.local_worker
+        if me in workers:
+            workers.remove(me)
+            workers = [me] + workers
         return workers
 
     def generate_shares(self, plain_text, number_of_shares=3):
         shares = []
+        plain_text.long()
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:48]: Q: In case plain_text is a ```float``` this will simply do a "cast" to an ```int``` type, right?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 9]
```diff
@@ -79,63 +93,152 @@ def __map_modular_to_real(self, mod_number):
...
+    def __reshare(shares, workers):
+        """convert 3-out-of-3 secret sharing: {player i : share i}
+          to 2-out-of-3 sharing: {player i : (share i, share i+1)}  """
+        shares_map = {}
+        for i in range(len(shares)):
+            pointer = shares[(i + 1) % len(shares)].copy().move(workers[i])
+            shares_map[workers[i]] = (shares[i], pointer)
+        return shares_map
+
+    def conv2d(self, image, padding=0):
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:195]: Q: Should we have this in the ```RSS``` or in the ```Falcon``` class?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 10]
```diff
@@ -46,21 +61,20 @@ def __distribute_shares(workers, shares):
...
-        shares_map = self.get_shares_map()
-        shares = self.__retrieve_shares(shares_map)
+        shares = self.retrieve_shares()
         plain_text_mod = self.__sum_shares(shares)
         plain_text = self.__map_modular_to_real(plain_text_mod)
         return plain_text
 
-    def __retrieve_shares(self, shares_map):
-        pointers = self.__retrieve_pointers(shares_map)
+    def retrieve_shares(self):
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:69]: Q: Can we make this ```__retrieve_shares```?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 11]
```diff
@@ -46,21 +61,20 @@ def __distribute_shares(workers, shares):
...
+        pointers = self.retrieve_pointers()
         shares = []
         for pointer in pointers:
             shares.append(pointer.get())
         return shares
 
-    @staticmethod
-    def __retrieve_pointers(shares_map):
+    def retrieve_pointers(self):
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:76]: Q: Also this one, ``` __retrieve_pointers?```

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 12]
```diff
@@ -146,6 +249,30 @@ def get_shares_map(self):
         """
         return self.child
 
+    def __set_shares_map(self, shares_map):
+        self.child = shares_map
+        return self
+
+    def apply_to_shares(self, function, *args, **kwargs):
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:256]: Q: This is used only internally. Should it be ```public```?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 5 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:33]: No, not anymore, this function was used to kinda init the object but this doesn't happen anymore

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 6 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:14]: this is checked in the __generate_shares_map function. if plain_text's None it returns None

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 7 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:18]: sometimes you need to make an empty tensor like an empty torch tensor, it should be acceptable I think

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 8 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:48]: it will cast it from any torch.dtype to torch.long

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 9 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:195]: hmm, it could be implemented both ways, but the paper described it as an operation of the sharing scheme, so let's stick to the paper for now, it can be changed easily later if it is inconvenient

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 10 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:69]: I meant to make them public since they are very useful for debugging to see the shares before adding them, it's basically the reconstruction algorithm but you'll have to add the shares manually

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 12 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:256]: any reshaping operation like view, unfold, reshape (or any other non-arithmetic operation) will use this function, so it will be used externally

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 6 above)
Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:12]: Hmm...seems ok - when refactoring we should stick with the same name method (we have ```share``` for ```AST``` and we do the sharing in the constructor for the ```RST```)

**Review by gmuraru — APPROVED** [review]

**Comment by gmuraru** [comment]
LGTM! After you merge it, could you give me a ping to refactor [this](https://github.com/OpenMined/PySyft/pull/4474)
